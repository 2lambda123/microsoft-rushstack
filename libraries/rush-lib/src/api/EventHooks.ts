// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { IEventHooksJson } from './RushConfiguration';
import { Enum } from '@rushstack/node-core-library';

/**
 * Events happen during Rush runs.
 * @beta
 */
// eslint-disable-next-line @typescript-eslint/typedef
export const Event = {
  /**
   * Pre Rush install event
   */
  preRushInstall: 'preRushInstall',
  /**
   * Post Rush install event
   */
  postRushInstall: 'postRushInstall',
  /**
   * Pre Rush build event
   */
  preRushBuild: 'preRushBuild',
  /**
   * Post Rush build event
   */
  postRushBuild: 'postRushBuild'
} as const;
export type Event = typeof Event[keyof typeof Event];

/**
 * This class represents Rush event hooks configured for this repo.
 * Hooks are customized script actions that Rush executes when specific events occur.
 * The actions are expressed as a command-line that is executed using the operating system shell.
 * @beta
 */
export class EventHooks {
  private _hooks: Map<Event, string[]>;

  /**
   * @internal
   */
  public constructor(eventHooksJson: IEventHooksJson) {
    this._hooks = new Map<Event, string[]>();
    for (const [name, eventHooks] of Object.entries(eventHooksJson)) {
      const eventName: Event | undefined = Enum.tryGetValueByKey(Event, name);
      if (eventName) {
        this._hooks.set(eventName, [...eventHooks] || []);
      }
    }
  }

  /**
   * Return all the scripts associated with the specified event.
   * @param event - Rush event
   */
  public get(event: Event): string[] {
    return this._hooks.get(event) || [];
  }
}
