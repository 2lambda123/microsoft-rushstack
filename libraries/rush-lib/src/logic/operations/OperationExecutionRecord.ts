// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import {
  DiscardStdoutTransform,
  SplitterTransform,
  StderrLineTransform,
  StdioSummarizer,
  TerminalWritable,
  TextRewriterTransform
} from '@rushstack/terminal';
import { InternalError, ITerminal, NewlineKind, Terminal } from '@rushstack/node-core-library';
import { CollatedTerminal, CollatedWriter, StreamCollator } from '@rushstack/stream-collator';

import { OperationStatus } from './OperationStatus';
import { IOperationRunner, IOperationRunnerContext } from './IOperationRunner';
import { Operation } from './Operation';
import { Stopwatch } from '../../utilities/Stopwatch';
import { OperationStateFile } from './OperationStateFile';
import { IOperationProcessor } from './IOperationProcessor';
import { CollatedTerminalProvider } from '../../utilities/CollatedTerminalProvider';
import { IOperationHashes } from './OperationHash';

export interface IOperationExecutionRecordContext {
  streamCollator: StreamCollator;

  debugMode: boolean;
  quietMode: boolean;
}

/**
 * Internal class representing everything about executing an operation
 */
export class OperationExecutionRecord implements IOperationRunnerContext {
  /**
   * The current execution status of an operation. Operations start in the 'ready' state,
   * but can be 'blocked' if an upstream operation failed. It is 'executing' when
   * the operation is executing. Once execution is complete, it is either 'success' or
   * 'failure'.
   */
  public status: OperationStatus = OperationStatus.Ready;

  /**
   * The error which occurred while executing this operation, this is stored in case we need
   * it later (for example to re-print errors at end of execution).
   */
  public error: Error | undefined = undefined;

  /**
   * This number represents how far away this Operation is from the furthest "root" operation (i.e.
   * an operation with no consumers). This helps us to calculate the critical path (i.e. the
   * longest chain of projects which must be executed in order, thereby limiting execution speed
   * of the entire operation tree.
   *
   * This number is calculated via a memoized depth-first search, and when choosing the next
   * operation to execute, the operation with the highest criticalPathLength is chosen.
   *
   * Example:
   *        (0) A
   *             \
   *          (1) B     C (0)         (applications)
   *               \   /|\
   *                \ / | \
   *             (2) D  |  X (1)      (utilities)
   *                    | / \
   *                    |/   \
   *                (2) Y     Z (2)   (other utilities)
   *
   * All roots (A & C) have a criticalPathLength of 0.
   * B has a score of 1, since A depends on it.
   * D has a score of 2, since we look at the longest chain (e.g D->B->A is longer than D->C)
   * X has a score of 1, since the only package which depends on it is A
   * Z has a score of 2, since only X depends on it, and X has a score of 1
   * Y has a score of 2, since the chain Y->X->C is longer than Y->C
   *
   * The algorithm is implemented in AsyncOperationQueue.ts as calculateCriticalPathLength()
   */
  public criticalPathLength: number | undefined = undefined;

  public silent: boolean = false;

  public isSkipAllowed: boolean = true;
  public isCacheReadAllowed: boolean = true;
  public isCacheWriteAllowed: boolean = true;

  public hashes: IOperationHashes | undefined = undefined;

  /**
   * The set of operations that must complete before this operation executes.
   */
  public readonly dependencies: Set<OperationExecutionRecord> = new Set();
  /**
   * The set of operations that depend on this operation.
   */
  public readonly consumers: Set<OperationExecutionRecord> = new Set();

  public readonly operation: Operation;

  public readonly stopwatch: Stopwatch = new Stopwatch();
  public readonly stdioSummarizer: StdioSummarizer = new StdioSummarizer();

  public runner: IOperationRunner;
  public readonly weight: number;
  public readonly _operationStateFile: OperationStateFile | undefined;

  private readonly _context: IOperationExecutionRecordContext;

  private _writer: CollatedWriter | undefined = undefined;
  private _terminalWritable: TerminalWritable | undefined = undefined;
  private _terminal: ITerminal | undefined = undefined;

  public constructor(operation: Operation, context: IOperationExecutionRecordContext) {
    this.operation = operation;
    const { runner } = operation;

    if (!runner) {
      throw new InternalError(
        `Operation for phase '${operation.associatedPhase.name}' and project '${operation.associatedProject.packageName}' has no runner.`
      );
    }
    this.silent = runner.silent;

    this.runner = runner;
    this.weight = operation.weight;

    const { associatedProject } = operation;

    this._operationStateFile = associatedProject
      ? new OperationStateFile({
          filename: `${associatedProject.projectFolder}/${operation.metadataFolderRelativePath}/state.json`
        })
      : undefined;
    this._context = context;
  }

  public get name(): string {
    return this.runner.name;
  }

  public get debugMode(): boolean {
    return this._context.debugMode;
  }

  public get quietMode(): boolean {
    return this._context.quietMode;
  }

  public get nonCachedDurationMs(): number | undefined {
    // Lazy calculated because the state file is created/restored later on
    return this._operationStateFile?.state?.nonCachedDurationMs;
  }

  public get terminalWritable(): TerminalWritable {
    if (!this._terminalWritable) {
      const stderrLineTransform: StderrLineTransform = new StderrLineTransform({
        destination: this.stdioSummarizer,
        newlineKind: NewlineKind.Lf // for StdioSummarizer
      });

      const discardTransform: DiscardStdoutTransform = new DiscardStdoutTransform({
        destination: this._collatedWriter
      });

      const splitterTransform: SplitterTransform = new SplitterTransform({
        destinations: [this.quietMode ? discardTransform : this._collatedWriter, stderrLineTransform]
      });

      const normalizeNewlineTransform: TextRewriterTransform = new TextRewriterTransform({
        destination: splitterTransform,
        normalizeNewlines: NewlineKind.Lf,
        ensureNewlineAtEnd: true
      });

      this._terminalWritable = normalizeNewlineTransform;
    }
    return this._terminalWritable;
  }

  public get terminal(): ITerminal {
    if (!this._terminal) {
      const collatedTerminal: CollatedTerminal = new CollatedTerminal(this.terminalWritable);
      this._terminal = new Terminal(new CollatedTerminalProvider(collatedTerminal));
    }
    return this._terminal;
  }

  private get _collatedWriter(): CollatedWriter {
    // Lazy instantiate because the registerTask() call affects display ordering
    if (!this._writer) {
      this._writer = this._context.streamCollator.registerTask(this.name);
    }
    return this._writer;
  }

  public async executeAsync(onResult: (record: OperationExecutionRecord) => void): Promise<void> {
    this.status = OperationStatus.Executing;

    try {
      let status: OperationStatus = this.status;
      const processor: IOperationProcessor | undefined = this.operation.processor;
      if (processor) {
        // Handle, e.g. build cache read
        status = await processor.beforeBuildAsync(this);
        if (status !== OperationStatus.Ready) {
          this.status = status;
          await this._operationStateFile?.tryRestoreAsync();
          return onResult(this);
        }
      }

      this.stopwatch.start();
      status = await this.runner.executeAsync(this);
      this.stopwatch.stop();

      await this._operationStateFile?.writeAsync({
        nonCachedDurationMs: this.stopwatch.duration * 1000,
        hashes: this.hashes,
        status
      });

      if (processor) {
        // Handle, e.g. build cache write
        status = await processor.afterBuildAsync(this, status);
      }

      this.status = status;
      // Delegate global state reporting
      onResult(this);
    } catch (error) {
      this.status = OperationStatus.Failure;
      this.error = error;
      // Delegate global state reporting
      onResult(this);
    } finally {
      this.silent = this.runner.silent;
      this._terminalWritable?.close();
      this._writer?.close();
      this.stopwatch.stop();
    }
  }
}
