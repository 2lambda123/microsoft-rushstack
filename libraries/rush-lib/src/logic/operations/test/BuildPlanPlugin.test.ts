// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

jest.mock('@rushstack/terminal', () => {
  const originalModule = jest.requireActual('@rushstack/terminal');
  return {
    ...originalModule,
    ConsoleTerminalProvider: {
      ...originalModule.ConsoleTerminalProvider,
      supportsColor: true
    }
  };
});
import { MockWritable, Terminal } from '@rushstack/terminal';
import { BuildPlanPlugin } from '../BuildPlanPlugin';
import {
  type ICreateOperationsContext,
  IExecuteOperationsContext,
  PhasedCommandHooks
} from '../../../pluginFramework/PhasedCommandHooks';
import { CollatedTerminalProvider } from '../../../utilities/CollatedTerminalProvider';
import { CollatedTerminal, StreamCollator } from '@rushstack/stream-collator';
import type { Operation } from '../Operation';
import path from 'path';
import { RushConfiguration } from '../../../api/RushConfiguration';
import {
  CommandLineConfiguration,
  type IPhase,
  type IPhasedCommandConfig
} from '../../../api/CommandLineConfiguration';
import type { ICommandLineJson } from '../../../api/CommandLineJson';
import { JsonFile } from '@rushstack/node-core-library';
import { OperationExecutionRecord } from '../OperationExecutionRecord';
import { PhasedOperationPlugin } from '../PhasedOperationPlugin';
import type { RushConfigurationProject } from '../../../api/RushConfigurationProject';
import { RushConstants } from '../../RushConstants';
import { MockOperationRunner } from './MockOperationRunner';
import { ProjectChangeAnalyzer } from '../../ProjectChangeAnalyzer';

const mockWritable: MockWritable = new MockWritable();
const mockTerminal: Terminal = new Terminal(new CollatedTerminalProvider(new CollatedTerminal(mockWritable)));

const mockStreamWritable: MockWritable = new MockWritable();
const streamCollator = new StreamCollator({
  destination: mockStreamWritable
});

describe('BuildPlanPlugin', () => {
  const rushJsonFile: string = path.resolve(__dirname, `../../test/workspaceRepo/rush.json`);
  const commandLineJsonFile: string = path.resolve(
    __dirname,
    `../../test/workspaceRepo/common/config/rush/command-line.json`
  );
  let rushConfiguration!: RushConfiguration;
  let commandLineConfiguration!: CommandLineConfiguration;

  beforeAll(() => {
    rushConfiguration = RushConfiguration.loadFromConfigurationFile(rushJsonFile);
    const commandLineJson: ICommandLineJson = JsonFile.load(commandLineJsonFile);

    commandLineConfiguration = new CommandLineConfiguration(commandLineJson);
  });

  function createMockRunner(operations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
    for (const operation of operations) {
      const { associatedPhase, associatedProject } = operation;

      if (associatedPhase && associatedProject && !operation.runner) {
        const name: string = `${associatedProject.packageName} (${associatedPhase.name.slice(
          RushConstants.phaseNamePrefix.length
        )})`;

        operation.runner = new MockOperationRunner(name);
      }
    }

    return operations;
  }

  async function testCreateOperationsAsync(
    phaseSelection: Set<IPhase>,
    projectSelection: Set<RushConfigurationProject>,
    changedProjects: Set<RushConfigurationProject>
  ): Promise<Set<Operation>> {
    const hooks: PhasedCommandHooks = new PhasedCommandHooks();
    // Apply the plugin being tested
    new PhasedOperationPlugin().apply(hooks);
    // Add mock runners for included operations.
    hooks.createOperations.tap('MockOperationRunnerPlugin', createMockRunner);

    const context: Pick<
      ICreateOperationsContext,
      'phaseOriginal' | 'phaseSelection' | 'projectSelection' | 'projectsInUnknownState'
    > = {
      phaseOriginal: phaseSelection,
      phaseSelection,
      projectSelection,
      projectsInUnknownState: changedProjects
    };
    const operations: Set<Operation> = await hooks.createOperations.promise(
      new Set(),
      context as ICreateOperationsContext
    );

    return operations;
  }

  describe('build plan debugging', () => {
    it('should generate a build plan', async () => {
      const hooks: PhasedCommandHooks = new PhasedCommandHooks();

      new BuildPlanPlugin(mockTerminal).apply(hooks);
      const context: Pick<IExecuteOperationsContext, 'projectChangeAnalyzer' | 'projectConfigurations'> = {
        projectChangeAnalyzer: {
          [ProjectChangeAnalyzer.prototype._tryGetProjectDependenciesAsync.name]: async () => {
            return new Map();
          }
        } as unknown as ProjectChangeAnalyzer,
        projectConfigurations: new Map()
      };
      const buildCommand: IPhasedCommandConfig = commandLineConfiguration.commands.get(
        'build'
      )! as IPhasedCommandConfig;

      const operationMap = new Map();

      const operations = await testCreateOperationsAsync(
        buildCommand.phases,
        new Set(rushConfiguration.projects),
        new Set(rushConfiguration.projects)
      );
      operations.forEach((operation) => {
        operationMap.set(
          operation,
          new OperationExecutionRecord(operation, { debugMode: false, quietMode: true, streamCollator })
        );
      });

      await hooks.beforeExecuteOperations.promise(operationMap, context as IExecuteOperationsContext);

      const allOutput: string = mockWritable.getAllOutput();
      expect(allOutput).toMatchSnapshot();
    });
  });
});
