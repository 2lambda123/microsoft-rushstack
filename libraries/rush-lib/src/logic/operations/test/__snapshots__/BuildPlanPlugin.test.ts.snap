// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildPlanPlugin build plan debugging should generate a build plan 1`] = `"Build Plan Depth (deepest dependency tree): 5[n]Build Plan Width (maximum parallelism): 38[n]Number of Nodes per Depth: 22, 38, 33, 11, 1[n]Plan @ Depth 0 has 22 nodes and 0 dependents:[n]- a (no-deps)[n]- b (no-deps)[n]- c (no-deps)[n]- d (no-deps)[n]- e (no-deps)[n]- f (no-deps)[n]- g (no-deps)[n]- h (no-deps)[n]- a (upstream-1)[n]- a (upstream-2)[n]- a (upstream-1-self-upstream)[n]- i (no-deps)[n]- j (no-deps)[n]- i (upstream-1)[n]- j (upstream-1)[n]- i (upstream-2)[n]- j (upstream-2)[n]- a (upstream-3)[n]- i (upstream-3)[n]- j (upstream-3)[n]- i (upstream-1-self-upstream)[n]- j (upstream-1-self-upstream)[n]Plan @ Depth 1 has 38 nodes and 22 dependents:[n]- a (upstream-self)[n]- b (upstream-1)[n]- f (upstream-1)[n]- g (upstream-1)[n]- h (upstream-1)[n]- b (upstream-self)[n]- c (upstream-1)[n]- d (upstream-1)[n]- c (upstream-self)[n]- e (upstream-1)[n]- d (upstream-self)[n]- e (upstream-self)[n]- f (upstream-self)[n]- g (upstream-self)[n]- h (upstream-self)[n]- b (upstream-2)[n]- f (upstream-2)[n]- g (upstream-2)[n]- h (upstream-2)[n]- a (upstream-1-self)[n]- b (upstream-3)[n]- f (upstream-3)[n]- g (upstream-3)[n]- h (upstream-3)[n]- a (upstream-2-self)[n]- b (complex)[n]- f (complex)[n]- g (complex)[n]- h (complex)[n]- i (upstream-self)[n]- j (upstream-self)[n]- i (upstream-1-self)[n]- j (upstream-1-self)[n]- i (upstream-2-self)[n]- j (upstream-2-self)[n]- a (complex)[n]- i (complex)[n]- j (complex)[n]Plan @ Depth 2 has 33 nodes and 60 dependents:[n]- b (upstream-self)[n]- f (upstream-self)[n]- h (upstream-self)[n]- g (upstream-self)[n]- c (upstream-2)[n]- d (upstream-2)[n]- b (upstream-1-self)[n]- f (upstream-1-self)[n]- g (upstream-1-self)[n]- f (upstream-2)[n]- h (upstream-1-self)[n]- c (upstream-self)[n]- d (upstream-self)[n]- e (upstream-2)[n]- c (upstream-1-self)[n]- d (upstream-1-self)[n]- e (upstream-self)[n]- e (upstream-1-self)[n]- c (upstream-3)[n]- d (upstream-3)[n]- b (upstream-2-self)[n]- f (upstream-2-self)[n]- g (upstream-2-self)[n]- f (upstream-3)[n]- h (upstream-2-self)[n]- b (upstream-1-self-upstream)[n]- f (upstream-1-self-upstream)[n]- g (upstream-1-self-upstream)[n]- h (upstream-1-self-upstream)[n]- b (complex)[n]- f (complex)[n]- g (complex)[n]- h (complex)[n]Plan @ Depth 3 has 11 nodes and 93 dependents:[n]- e (upstream-3)[n]- c (upstream-2-self)[n]- d (upstream-2-self)[n]- c (upstream-1-self-upstream)[n]- d (upstream-1-self-upstream)[n]- f (upstream-1-self-upstream)[n]- e (upstream-2-self)[n]- e (upstream-1-self-upstream)[n]- c (complex)[n]- d (complex)[n]- f (complex)[n]Plan @ Depth 4 has 1 nodes and 104 dependents:[n]- e (complex)[n]##################################################[n]                  a (no-deps): (0)[n]                  b (no-deps): (0)[n]                  c (no-deps): (0)[n]                  d (no-deps): (0)[n]                  e (no-deps): (0)[n]                  f (no-deps): (0)[n]                  g (no-deps): (0)[n]                  h (no-deps): (0)[n]               a (upstream-1): (0)[n]               a (upstream-2): (0)[n] a (upstream-1-self-upstream): (0)[n]                  i (no-deps): (1)[n]                  j (no-deps): (2)[n]               i (upstream-1): (3)[n]               j (upstream-1): (4)[n]               i (upstream-2): (5)[n]               j (upstream-2): (6)[n]               a (upstream-3): (7)[n]               i (upstream-3): (8)[n]               j (upstream-3): (9)[n] i (upstream-1-self-upstream): (10)[n] j (upstream-1-self-upstream): (11)[n]            a (upstream-self): -(0)[n]               b (upstream-1): -(0)[n]               c (upstream-1): -(0)[n]               d (upstream-1): -(0)[n]               e (upstream-1): -(0)[n]               f (upstream-1): -(0)[n]               g (upstream-1): -(0)[n]               h (upstream-1): -(0)[n]               b (upstream-2): -(0)[n]               g (upstream-2): -(0)[n]               h (upstream-2): -(0)[n]               b (upstream-3): -(0)[n]               g (upstream-3): -(0)[n]               h (upstream-3): -(0)[n]          a (upstream-1-self): -(0)[n]          a (upstream-2-self): -(0)[n]            i (upstream-self): -(1)[n]            j (upstream-self): -(2)[n]          i (upstream-1-self): -(3)[n]          j (upstream-1-self): -(4)[n]          i (upstream-2-self): -(5)[n]          j (upstream-2-self): -(6)[n]                  a (complex): -(7)[n]                  i (complex): -(8)[n]                  j (complex): -(9)[n]            b (upstream-self): --(0)[n]            f (upstream-self): --(0)[n]            h (upstream-self): --(0)[n]            g (upstream-self): --(0)[n]               c (upstream-2): --(0)[n]               d (upstream-2): --(0)[n]               e (upstream-2): --(0)[n]               f (upstream-2): --(0)[n]               c (upstream-3): --(0)[n]               d (upstream-3): --(0)[n]               f (upstream-3): --(0)[n]          b (upstream-1-self): --(0)[n]          c (upstream-1-self): --(0)[n]          d (upstream-1-self): --(0)[n]          e (upstream-1-self): --(0)[n]          f (upstream-1-self): --(0)[n]          g (upstream-1-self): --(0)[n]          h (upstream-1-self): --(0)[n]          b (upstream-2-self): --(0)[n]          g (upstream-2-self): --(0)[n]          h (upstream-2-self): --(0)[n] b (upstream-1-self-upstream): --(0)[n] g (upstream-1-self-upstream): --(0)[n] h (upstream-1-self-upstream): --(0)[n]                  b (complex): --(0)[n]                  g (complex): --(0)[n]                  h (complex): --(0)[n]            c (upstream-self): ---(0)[n]            d (upstream-self): ---(0)[n]               e (upstream-3): ---(0)[n]          c (upstream-2-self): ---(0)[n]          d (upstream-2-self): ---(0)[n]          e (upstream-2-self): ---(0)[n]          f (upstream-2-self): ---(0)[n] c (upstream-1-self-upstream): ---(0)[n] d (upstream-1-self-upstream): ---(0)[n] e (upstream-1-self-upstream): ---(0)[n] f (upstream-1-self-upstream): ---(0)[n]                  c (complex): ---(0)[n]                  d (complex): ---(0)[n]                  f (complex): ---(0)[n]            e (upstream-self): ----(0)[n]                  e (complex): ----(0)[n]##################################################[n]Cluster 0:[n]- Dependencies: none[n]- Clustered by: [n]  - (a (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (d (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (e (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (f (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (g (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (d (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (e (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (f (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (g (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (d (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (e (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (f (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (g (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (a (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (b (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (d (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (e (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (c (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (f (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (g (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]  - (h (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: a (no-deps), b (no-deps), c (no-deps), d (no-deps), e (no-deps), f (no-deps), g (no-deps), h (no-deps), a (upstream-self), b (upstream-self), c (upstream-self), d (upstream-self), e (upstream-self), f (upstream-self), h (upstream-self), g (upstream-self), a (upstream-1), b (upstream-1), c (upstream-1), d (upstream-1), e (upstream-1), f (upstream-1), g (upstream-1), h (upstream-1), a (upstream-2), b (upstream-2), c (upstream-2), d (upstream-2), e (upstream-2), f (upstream-2), g (upstream-2), h (upstream-2), b (upstream-3), c (upstream-3), d (upstream-3), e (upstream-3), f (upstream-3), g (upstream-3), h (upstream-3), a (upstream-1-self), b (upstream-1-self), c (upstream-1-self), d (upstream-1-self), e (upstream-1-self), f (upstream-1-self), g (upstream-1-self), h (upstream-1-self), a (upstream-2-self), b (upstream-2-self), c (upstream-2-self), d (upstream-2-self), e (upstream-2-self), f (upstream-2-self), g (upstream-2-self), h (upstream-2-self), a (upstream-1-self-upstream), b (upstream-1-self-upstream), c (upstream-1-self-upstream), d (upstream-1-self-upstream), e (upstream-1-self-upstream), f (upstream-1-self-upstream), g (upstream-1-self-upstream), h (upstream-1-self-upstream), b (complex), c (complex), d (complex), e (complex), f (complex), g (complex), h (complex)[n]--------------------------------------------------[n]Cluster 1:[n]- Dependencies: none[n]- Clustered by: [n]  - (i (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: i (no-deps), i (upstream-self)[n]--------------------------------------------------[n]Cluster 2:[n]- Dependencies: none[n]- Clustered by: [n]  - (j (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: j (no-deps), j (upstream-self)[n]--------------------------------------------------[n]Cluster 3:[n]- Dependencies: none[n]- Clustered by: [n]  - (i (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: i (upstream-1), i (upstream-1-self)[n]--------------------------------------------------[n]Cluster 4:[n]- Dependencies: none[n]- Clustered by: [n]  - (j (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: j (upstream-1), j (upstream-1-self)[n]--------------------------------------------------[n]Cluster 5:[n]- Dependencies: none[n]- Clustered by: [n]  - (i (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: i (upstream-2), i (upstream-2-self)[n]--------------------------------------------------[n]Cluster 6:[n]- Dependencies: none[n]- Clustered by: [n]  - (j (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: j (upstream-2), j (upstream-2-self)[n]--------------------------------------------------[n]Cluster 7:[n]- Dependencies: none[n]- Clustered by: [n]  - (a (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: a (upstream-3), a (complex)[n]--------------------------------------------------[n]Cluster 8:[n]- Dependencies: none[n]- Clustered by: [n]  - (i (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: i (upstream-3), i (complex)[n]--------------------------------------------------[n]Cluster 9:[n]- Dependencies: none[n]- Clustered by: [n]  - (j (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"[n]- Operations: j (upstream-3), j (complex)[n]--------------------------------------------------[n]Cluster 10:[n]- Dependencies: none[n]- Operations: i (upstream-1-self-upstream)[n]--------------------------------------------------[n]Cluster 11:[n]- Dependencies: none[n]- Operations: j (upstream-1-self-upstream)[n]--------------------------------------------------[n]##################################################[n]"`;
